File: src/scrapio/resources/jobs.py
from __future__ import annotations
import asyncio
import time
from .._http import SyncHttpClient, AsyncHttpClient
from ..types import CreateJobRequest, Job, JobResult

TERMINAL = {"completed", "partial", "failed", "cancelled"}


class JobsResource:
    def __init__(self, http: SyncHttpClient) -> None:
        self._http = http

    def create(self, request: CreateJobRequest) -> Job:
        return self._http.request(
            "POST", "/v1/jobs",
            json=request.model_dump(exclude_none=True),
            response_model=Job,
        )

    def get(self, job_id: str) -> Job:
        return self._http.request("GET", f"/v1/jobs/{job_id}", response_model=Job)

    def get_result(self, job_id: str) -> JobResult:
        return self._http.request("GET", f"/v1/jobs/{job_id}/result", response_model=JobResult)

    def wait_for_completion(
        self,
        job_id: str,
        *,
        poll_interval: float = 2.0,
        timeout: float = 300.0,
    ) -> JobResult:
        deadline = time.monotonic() + timeout
        while time.monotonic() < deadline:
            job = self.get(job_id)
            if job.status in TERMINAL:
                return self.get_result(job_id)
            time.sleep(poll_interval)
        raise TimeoutError(f"Job {job_id} did not complete within {timeout}s")


class AsyncJobsResource:
    def __init__(self, http: AsyncHttpClient) -> None:
        self._http = http

    async def create(self, request: CreateJobRequest) -> Job:
        return await self._http.request(
            "POST", "/v1/jobs",
            json=request.model_dump(exclude_none=True),
            response_model=Job,
        )

    async def get(self, job_id: str) -> Job:
        return await self._http.request("GET", f"/v1/jobs/{job_id}", response_model=Job)

    async def get_result(self, job_id: str) -> JobResult:
        return await self._http.request("GET", f"/v1/jobs/{job_id}/result", response_model=JobResult)

    async def wait_for_completion(
        self,
        job_id: str,
        *,
        poll_interval: float = 2.0,
        timeout: float = 300.0,
    ) -> JobResult:
        deadline = asyncio.get_event_loop().time() + timeout
        while asyncio.get_event_loop().time() < deadline:
            job = await self.get(job_id)
            if job.status in TERMINAL:
                return await self.get_result(job_id)
            await asyncio.sleep(poll_interval)
        raise TimeoutError(f"Job {job_id} did not complete within {timeout}s")

File: jobs.go
package scrapio

import (
	"context"
	"fmt"
	"time"
)

type JobWebhook struct {
	URL string `json:"url"`
}

type CreateJobRequest struct {
	Kind    string         `json:"kind"`
	Input   map[string]any `json:"input"`
	Webhook *JobWebhook    `json:"webhook,omitempty"`
}

type Job struct {
	RequestID  string `json:"request_id"`
	JobID      string `json:"job_id"`
	JobType    string `json:"job_type"`
	Status     string `json:"status"`
	CreatedAt  string `json:"created_at"`
	UpdatedAt  string `json:"updated_at,omitempty"`
	WebhookURL string `json:"webhook_url,omitempty"`
}

type JobError struct {
	Code    string `json:"code"`
	Message string `json:"message"`
}

type JobResult struct {
	Job
	Result any       `json:"result,omitempty"`
	Error  *JobError `json:"error,omitempty"`
}

var terminalStatuses = map[string]bool{
	"completed": true,
	"partial":   true,
	"failed":    true,
	"cancelled": true,
}

type JobsResource struct{ h *httpClient }

func (r *JobsResource) Create(ctx context.Context, req *CreateJobRequest) (*Job, error) {
	var out Job
	return &out, r.h.post(ctx, "/v1/jobs", req, &out)
}

func (r *JobsResource) Get(ctx context.Context, jobID string) (*Job, error) {
	var out Job
	return &out, r.h.get(ctx, "/v1/jobs/"+jobID, nil, &out)
}

func (r *JobsResource) GetResult(ctx context.Context, jobID string) (*JobResult, error) {
	var out JobResult
	return &out, r.h.get(ctx, "/v1/jobs/"+jobID+"/result", nil, &out)
}

type WaitOptions struct {
	PollInterval time.Duration
	Timeout      time.Duration
}

func (r *JobsResource) WaitForCompletion(ctx context.Context, jobID string, opts *WaitOptions) (*JobResult, error) {
	interval := 2 * time.Second
	deadline := 5 * time.Minute
	if opts != nil {
		if opts.PollInterval > 0 {
			interval = opts.PollInterval
		}
		if opts.Timeout > 0 {
			deadline = opts.Timeout
		}
	}
	ctx, cancel := context.WithTimeout(ctx, deadline)
	defer cancel()

	for {
		job, err := r.Get(ctx, jobID)
		if err != nil {
			return nil, err
		}
		if terminalStatuses[job.Status] {
			return r.GetResult(ctx, jobID)
		}
		select {
		case <-ctx.Done():
			return nil, fmt.Errorf("scrapio: job %s did not complete within timeout", jobID)
		case <-time.After(interval):
		}
	}
}
